api/client/cp.go

```diff
@@ -1,57 +1,253 @@
-	cmd.Require(flag.Exact, 2)
+	cmd := cli.Subcmd(
+		"cp",
+		"CONTAINER:PATH LOCALPATH|-\nLOCALPATH|- CONTAINER:PATH",
+		"Copy files/folders between containers and the local filesystem.",
+		true,
+	)
 
+	cmd.Require(flag.Exact, 2)
 	cmd.ParseFlags(args, true)
 
-	// deal with path name with `:`
-	info := strings.SplitN(cmd.Arg(0), ":", 2)
+	if cmd.Arg(0) == "" {
+		return fmt.Errorf("must specify a source")
+	}
+	if cmd.Arg(1) == "" {
+		return fmt.Errorf("must specify a destination")
+	}
 
-	if len(info) != 2 {
-		return fmt.Errorf("Error: Path not specified")
+	srcContainer, srcPath := splitCpArg(cmd.Arg(0))
+	dstContainer, dstPath := splitCpArg(cmd.Arg(1))
+
+	var direction copyDirection
+	if srcContainer != "" {
+		direction |= fromContainer
+	}
+	if dstContainer != "" {
+		direction |= toContainer
 	}
 
-	cfg := &types.CopyConfig{
-		Resource: info[1],
+	switch direction {
+	case fromContainer:
+		return cli.copyFromContainer(srcContainer, srcPath, dstPath)
+	case toContainer:
+		return cli.copyToContainer(srcPath, dstContainer, dstPath)
+	case acrossContainers:
+		// Copying between containers isn't supported yet.
+		return fmt.Errorf("copying between containers is not yet supported")
+	default:
+		// User didn't specify any container.
+		return fmt.Errorf("must specify at least one container source")
 	}
-	stream, statusCode, err := cli.call("POST", "/containers/"+info[0]+"/copy", cfg, nil)
-	if stream != nil {
-		defer stream.Close()
+}
+
+// We use `:` as a delimiter between CONTAINER and PATH, but `:` could also be
+// in a valid LOCALPATH, like `file:name.txt`. We can resolve this ambiguity by
+// requiring a LOCALPATH with a `:` to be made explicit with a relative or
+// absolute path:
+// 	`/path/to/file:name.txt` or `./file:name.txt`
+//
+// This is apparently how `scp` handles this as well:
+// 	http://www.cyberciti.biz/faq/rsync-scp-file-name-with-colon-punctuation-in-it/
+//
+// We can't simply check for a filepath separator because container names may
+// have a separator, e.g., "host0/cname1" if container is in a Docker cluster,
+// so we have to check for a `/` or `.` prefix. Also, in the case of a Windows
+// client, a `:` could be part of an absolute Windows path, in which case it
+// is immediately proceeded by a backslash.
+func splitCpArg(arg string) (container, path string) {
+	if filepath.IsAbs(arg) {
+		// Explicit local absolute path, e.g., `C:\foo` or `/foo`.
+		return "", arg
 	}
-	if statusCode == 404 {
-		return fmt.Errorf("No such container: %v", info[0])
+
+	parts := strings.SplitN(arg, ":", 2)
+
+	switch {
+	case len(parts) == 1:
+		// No `:` in the arg.
+		fallthrough
+	case strings.HasPrefix(parts[0], "."):
+		// Explicit local relative path.
+		return "", arg
+	default:
+		return parts[0], parts[1]
+	}
+}
+
+type containerPathStat struct {
```

Comment:
You should probably move this into api/types/types.go and then you can use the same exact struct on the cli and in the container code - removing "ContainerResourceStat"